### ISSUE TITLE
[Feature request] YAML headers support

### ISSUE THREAD

**Issue by samarulmeu** [SidOfc/mkdx#96]
Half of the time I am using markdown for writing blog posts (I use a Static Site Generator) and the plugin I used before (vim-pandoc-syntax) supported in a very nice way YAML headers highlighting. Actually, for me, was the only missing feature of this plugin as mkdx does it better for all the others (italic, bold, headers). Ok, maybe also the quote highlighting could be improve (to characters have the same color as the `>` mark). 

I think it will be great if mkdx will also support this kind of highlighting.

**Comment by SidOfc** [comment]
Actually mkdx does not try to be a syntax highlighting plugin, it relies on your regular markdown highlighter for most of the heavy lifting.

Some highlighters create inconsistent results and could use "some patches", this is why mkdx has some highlighting functionality, some "patched" things include:

- List item highlighting, ordered and unordered
- Checkbox state (done / pending / unstarted) highlighting
- Block code snippets
- "Bold + Italic" highlighting specifically
- Tables
- keyboard tags `<kbd>1</kbd>+<kbd>2</kbd>`
- strike text

All these improved highlights are tied to the fact that mkdx has functionality which would be nicer to use if highlighting didn't "suck" by default in some scenario's (disclaimer: it's hard, and I very much respect those that write these syntax files btw, not trying to stab anyone at all).

With that being said, I'm not even sure if it's hard to add. I want to understand a bit more about why it's wanted, just colors? Is it visually disturbing using default markdown highlighting?

**Comment by samarulmeu** [comment]
I think I should have provided an example, so here it is:

![yamlHeader](https://user-images.githubusercontent.com/39216570/76874949-8d0c9e00-6878-11ea-8f3d-f159b6398b4b.png)

Yes, it is mostly about color rendering. Using field colors it is easier to handle YAML headers. 

If it is too much trouble, you can close the issue. In a way it is a particular request.

**Comment by SidOfc** [comment]
I took a screenshot of mine now too: 

![image](https://user-images.githubusercontent.com/3225058/76885092-01e2d680-687f-11ea-97ef-eb5dbc98d36c.png)

In my case, it's the last line which is actually bad because if I add something below:

![image](https://user-images.githubusercontent.com/3225058/76885157-17f09700-687f-11ea-8e8c-fc12d77b66a5.png)

This is because the default markdown syntax highlighter doesn't understand YAML front-matter it seems, and coincidentally the highlighting I see is actually that of a H2 title:

![image](https://user-images.githubusercontent.com/3225058/76885345-60a85000-687f-11ea-9804-6dbf23c5b2b3.png)

That said I am not one that knows a lot about doing actual syntax highlighting, I've modified existing stuff before, created some "simple" things, but here we want to basically "parse any YAML" when it is enclosed by `---` lines AND at the top of the file.

I do think it is possible to load the YAML syntax highlighter for specific sections of text though, maybe it is something for me to look into in the long run since I assume multiple users will want to have decent front-matter highlighting.

If you manage to modify your front-matter syntax I'd be interested in knowing what you did (add a plugin, some custom vimrc stuff etc) so that I may try to reproduce it in mkdx. In the meantime I'm going to leave this issue open as there could be more interest in this.

(note to self: perhaps I should attempt to write my own markdown syntax file once)

**Comment by samarulmeu** [comment]
The syntax highlightning from my example it is done by the *vim-pandoc-syntax* plugin I was mentioning.

Maybe I should also look and see if I can understand how it is done and help you implement it.

**Comment by SidOfc** [comment]
Keep me posted! I'm working on the list of issues bit by bit, getting there :)

**Comment by samarulmeu** [comment]
So, I think that it will be easier that I thought. The plugin I mentioned (*vim-pandoc-syntax*) plugin wasn't doing any highlighting for yaml files. This, **vim** already knows how to do it in the file `/usr/share/vim/vim81/syntax/yaml.vim`. And you can see that if you open, only with _mkdx_ enabled, a `test.yaml` file (see the extention) that  contains the next lines

```
---
layout: post
title: "Post title"
description: "Post description"
date: 2020-03-17 14:39:57+0200
permalink: /some-link/
image: 2020/03/image.jpg
categories: [category]
tags: [tag1, tag2]
comments: true
comments_locked: false
---
```

you will get a nice highlighting. 

The only think, and this I did not figure it out, is to tell the plugin that all text between `---` lines is a YAML header and treat it properly. 

Hope my explanation was clear enough and I helped a little.

**Comment by SidOfc** [comment]
@samarulmeu certainly!

I know how to do it, kinda, because the yaml syntax file defines syntax groups, I can choose to only load those "within" a specific `mkdxYaml` group for example. I am not sure how exactly it works but figured I'd steal from another plugin / syntax file since I believe some markdown plugins do it for highlighting the languages in marked fenced code blocks.

Cheers so far :+1:

**Comment by samarulmeu** [comment]
So, after searching and reading the code from the above mentioned plugin I got to this lines:

```
syn include @YAML syntax/yaml.vim
syn region YAMLHeader start=/\%(\%^\|\_^\s*\n\)\@<=\_^-\{3}\ze\n.\+/ end=/^\([-.]\)\1\{2}$/ keepend contains=@YAML containedin=TOP
```

Maybe they are of some help.

**Comment by SidOfc** [comment]
@samarulmeu I quite literally copied them and it worked :raised_hands: 

I also added a check to see if the YAML syntax file exists. If this isn't done, vim could crash with an ugly warning on startup. Thanks for the research and essentially the implementation, I think you've made more people a bit happier when they look at their front-matter next time. :+1:

**Comment by samarulmeu** [comment]
I hope that I was clear enough. The line is from the vim-pandoc-syntax plugin. I think they have a MIT license, but I think they should have credit for it. I look tomorow at the exact line and file.

**Comment by SidOfc** [comment]
@samarulmeu I'll be happy to include a line of credit for whomever wrote this piece of code, I might beat you to it :)

**Comment by SidOfc** [comment]
As a matter of fact, it doesn't seem to be the case since I don't think that "two lines" falls under "substantial parts" of the software. However I will add a line of credit above the code (I've done this for other snippets I shamelessly stole too heh)

:+1:

**Comment by SidOfc** [comment]
Alright, credits added, I looked a bit closer and also decided to include the try catch :)

Also fixed another bug, it seems that not all highlighting was shown before. Should all work properly now though. Keep me posted :)

**Comment by samarulmeu** [comment]
I checked and everything is perfect now and the new Bold and Italic is looking nice (taking the color from the theme). 

I looked again in the code from vim-pandoc, and maybe you were right, actually the only thing borrowed was the regular expression, all other stuffs are common knowledge. Sometimes I am more catholic than the pope :). 

Thank you again! Now I have no more reasons to procrastinate and start writing. 

Mkdx is a great plugin and I will advertise it. 

And closing...
